### ISSUE TITLE
pip install --no-deps --no-binary :all: ipykernel fails

### ISSUE THREAD

**Issue by cglong** [ipython/ipykernel#250]
I'm trying to add IPyKernel to [Homebrew](https://brew.sh/), a package manager for macOS. [Homebrew manages Python applications](http://docs.brew.sh/Python-for-Formula-Authors.html) by downloading the source tarballs for all dependent modules, then installs the tarball in a new virtualenv with `pip install --no-deps --no-binary :all: --ignore-installed`. Other modules install fine with this, but IPyKernel fails with the following output.

I can also create a similar, but slightly different repro, by creating my own virtualenv outside the Homebrew infrastructure and calling `pip install --no-deps --no-binary :all: ipykernel`.

It looks like `setup.py` expects IPython to be available at build-time, rather than runtime. Is there a way we can update the script to fallback gracefully in this case?
```
/usr/local/Cellar/ipython@5/5.4.1/libexec/bin/pip
install
-v
--no-deps
--no-binary
:all:
--ignore-installed
/private/tmp/ipython@5--ipykernel-20170613-20257-xee6ki/ipykernel-4.6.1

Processing /private/tmp/ipython@5--ipykernel-20170613-20257-xee6ki/ipykernel-4.6.1
  Running setup.py (path:/private/tmp/pip-3F1tdh-build/setup.py) egg_info for package from file:///private/tmp/ipython%405--ipykernel-20170613-20257-xee6ki/ipykernel-4.6.1
    Running command python setup.py egg_info
    running egg_info
    creating pip-egg-info/ipykernel.egg-info
    writing requirements to pip-egg-info/ipykernel.egg-info/requires.txt
    writing pip-egg-info/ipykernel.egg-info/PKG-INFO
    writing top-level names to pip-egg-info/ipykernel.egg-info/top_level.txt
    writing dependency_links to pip-egg-info/ipykernel.egg-info/dependency_links.txt
    writing manifest file 'pip-egg-info/ipykernel.egg-info/SOURCES.txt'
    reading manifest file 'pip-egg-info/ipykernel.egg-info/SOURCES.txt'
    writing manifest file 'pip-egg-info/ipykernel.egg-info/SOURCES.txt'
  Source in /private/tmp/pip-3F1tdh-build has version 4.6.1, which satisfies requirement ipykernel==4.6.1 from file:///private/tmp/ipython%405--ipykernel-20170613-20257-xee6ki/ipykernel-4.6.1
Skipping bdist_wheel for ipykernel, due to binaries being disabled for it.
Installing collected packages: ipykernel
  Running setup.py install for ipykernel: started
    Running command /usr/local/Cellar/ipython@5/5.4.1/libexec/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-3F1tdh-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-CMBJEz-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/ipython@5/5.4.1/libexec/bin/../include/site/python2.7/ipykernel
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/tmp/pip-3F1tdh-build/setup.py", line 91, in <module>
        from ipykernel.kernelspec import write_kernel_spec, make_ipkernel_cmd, KERNEL_NAME
      File "ipykernel/__init__.py", line 2, in <module>
        from .connect import *
      File "ipykernel/connect.py", line 13, in <module>
        from IPython.core.profiledir import ProfileDir
    ImportError: No module named IPython.core.profiledir
    Running setup.py install for ipykernel: finished with status 'error'
Cleaning up...
  Removing source in /private/tmp/pip-3F1tdh-build
Command "/usr/local/Cellar/ipython@5/5.4.1/libexec/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-3F1tdh-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-CMBJEz-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/ipython@5/5.4.1/libexec/bin/../include/site/python2.7/ipykernel" failed with error code 1 in /private/tmp/pip-3F1tdh-build/
Exception information:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/local/lib/python2.7/site-packages/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/usr/local/lib/python2.7/site-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/usr/local/lib/python2.7/site-packages/pip/req/req_install.py", line 878, in install
    spinner=spinner,
  File "/usr/local/lib/python2.7/site-packages/pip/utils/__init__.py", line 707, in call_subprocess
    % (command_desc, proc.returncode, cwd))
InstallationError: Command "/usr/local/Cellar/ipython@5/5.4.1/libexec/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-3F1tdh-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-CMBJEz-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/ipython@5/5.4.1/libexec/bin/../include/site/python2.7/ipykernel" failed with error code 1 in /private/tmp/pip-3F1tdh-build/
```
Output of `IPython.sys_info()` as requested:
```
{'commit_hash': u'1149d1700',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/usr/local/Cellar/ipython@5/5.4.1/libexec/lib/python2.7/site-packages/IPython',
 'ipython_version': '5.4.1',
 'os_name': 'posix',
 'platform': 'Darwin-16.5.0-x86_64-i386-64bit',
 'sys_executable': '/usr/local/Cellar/ipython@5/5.4.1/libexec/bin/python2.7',
 'sys_platform': 'darwin',
 'sys_version': '2.7.13 (default, Dec 18 2016, 07:03:39) \n[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]'}
```

**Comment by cglong** [comment]
Any ideas here? 😄

**Comment by Carreau** [comment]
Can you apply a patch with homebrew before building ? The culprit is 
```
# setup.py
if any(a.startswith(('bdist', 'build', 'install')) for a in sys.argv):
    from ipykernel.kernelspec import write_kernel_spec, make_ipkernel_cmd, KERNEL_NAME

    argv = make_ipkernel_cmd(executable='python')
    dest = os.path.join(here, 'data_kernelspec')
    if os.path.exists(dest):
        shutil.rmtree(dest)
    write_kernel_spec(dest, overrides={'argv': argv})

    setup_args['data_files'] = [
        (pjoin('share', 'jupyter', 'kernels', KERNEL_NAME), glob(pjoin(dest, '*'))),
    ]

```

This is because ipykernel need some functionality of IPython to actually build correctly. So Either you comment it and replace that, or you need to have IPython installed in the venv before installing ipykernel. You could just comment the above in setup.py but then the Jupyter notebook is likely to not find the Python kernel. 

Can't you also try to install from the wheel ? That should work and should not be any different.

**Comment by cglong** [comment]
Thanks for your help @Carreau; I was able to get it working as expected! See below for more detailed answers to your questions.

> Can you apply a patch with homebrew before building ?

Yes, we could apply a patch, but patches in Homebrew are expected to be temporary and ultimately fixed upstream.

> Can't you also try to install from the wheel ?

Homebrew has a policy that [all formulae be buildable from source](http://docs.brew.sh/Acceptable-Formulae.html).

> or you need to have IPython installed in the venv before installing ipykernel.

This was the missing piece! IPython was being installed as the last step, so by completing the installation of IPython and all its dependencies before beginning the installation of IPyKernel and its dependencies, I was able to open Homebrew/homebrew-core#14916. Thanks again for your help! 😄
